src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1423,23 +1423,28 @@ mongoc_uri_parse (mongoc_uri_t *uri, const char *str, bson_error_t *error)
       goto error;
    }
 
-   before_slash = scan_to_unichar (str, '/', "", &tmp);
-   if (!before_slash) {
-      before_slash = bson_strdup (str);
-      str += strlen (before_slash);
+   // Check for delimiting slash (not required)
+   before_options = scan_to_unichar (str, '/', "", &tmp);
+   if (!before_options) {
+      before_options = scan_to_unichar (str, '?', "", &tmp);
+   }
+
+   if (!before_options) {
+      before_options = bson_strdup (str);
+      str += strlen (before_options);
    } else {
       str = tmp;
    }
 
-   if (!mongoc_uri_parse_before_slash (uri, before_slash, error)) {
+   if (!mongoc_uri_parse_before_slash (uri, before_options, error)) {
       goto error;
    }
 
    BSON_ASSERT (str);
 
    if (*str) {
-      if (*str == '/') {
-         str++;
+      // Check for valid end of hostname delimeter (skip slash if necessary)
+      if (*str == '?' || *str++ == '/') {
```

Comment:
Suggest splitting the conditions for simplicity. IMO the conditionally run `*str++` adds some difficulty to reason about. Example:

```c
if (*str != '/' && *str != '?') {
   MONGOC_URI_ERROR (error, "%s", "Expected end of hostname delimiter");
   goto error;
}

if (*str == '/') {
   // Try to parse database.
   str++;
   if (*str) {
      if (!mongoc_uri_parse_database (uri, str, &str)) {
         MONGOC_URI_ERROR (error, "%s", "Invalid database name in URI");
         goto error;
      }
   }
}

if (*str == '?') {
   // Try to parse options.
   str++;
   if (*str) {
      if (!mongoc_uri_parse_options (uri, str, false /* from DNS */, error)) {
         goto error;
      }
   }
}
```